File: backend/app/schemas/investigation.py
from pydantic import BaseModel, Field


class QueryFilters(BaseModel):
    preset: str = Field(default="all", pattern="^(all|correct|incorrect|false_positive|false_negative)$")
    ground_truth: str | None = None
    prediction: str | None = None
    confidence_min: float | None = Field(default=None, ge=0.0, le=1.0)
    confidence_max: float | None = Field(default=None, ge=0.0, le=1.0)
    sample_id: str | int | None = None
    limit: int = Field(default=100, ge=1, le=1000)
    offset: int = Field(default=0, ge=0)


class QueryResult(BaseModel):
    rows: list[dict]
    total: int
    offset: int
    limit: int
    columns: list[str]


class Insights(BaseModel):
    accuracy: float | None = None
    total_rows: int = 0
    correct_count: int = 0
    incorrect_count: int = 0
    false_positive_count: int = 0
    false_negative_count: int = 0
    prediction_distribution: dict[str, int] = {}
    ground_truth_distribution: dict[str, int] = {}
    top_confusion_pairs: list[dict] = []


class RowDetail(BaseModel):
    evaluation_row: dict
    dataset_row: dict | None = None

File: frontend/src/api/investigation.ts
import api from "./client";

export interface QueryFilters {
  preset?: string;
  ground_truth?: string;
  prediction?: string;
  confidence_min?: number;
  confidence_max?: number;
  sample_id?: string | number;
  limit?: number;
  offset?: number;
}

export interface QueryResult {
  rows: Array<Record<string, any>>;
  total: number;
  offset: number;
  limit: number;
  columns: string[];
}

export interface Insights {
  accuracy: number | null;
  total_rows: number;
  correct_count: number;
  incorrect_count: number;
  false_positive_count: number;
  false_negative_count: number;
  prediction_distribution: Record<string, number>;
  ground_truth_distribution: Record<string, number>;
  top_confusion_pairs: Array<Record<string, any>>;
}

export interface RowDetail {
  evaluation_row: Record<string, any>;
  dataset_row: Record<string, any> | null;
}

export interface QueryFiltersPayload extends Omit<QueryFilters, "limit" | "offset"> {
  limit?: number;
  offset?: number;
}

export const queryEvaluation = (runId: number, filters: QueryFiltersPayload) =>
  api.post<QueryResult>(`/api/runs/${runId}/investigation/query`, filters).then((r) => r.data);

export const getRowDetail = (runId: number, index: number) =>
  api.post<RowDetail>(`/api/runs/${runId}/investigation/row/${index}`).then((r) => r.data);

export const computeInsights = (runId: number) =>
  api.post<Insights>(`/api/runs/${runId}/investigation/insights`).then((r) => r.data);
